### PR TITLE
feat(node-cpu-hog): Adding the node-cpu-hog experiment

### PR DESCRIPTION
**Pull request by ispeakc0de**
Signed-off-by: shubhamchaudhary <shubham.chaudhary@mayadata.io>

- Adding the node-cpu-hog experiment

- fixes: litmuschaos/litmus#1670

### CODE REVIEW

**Comment by ispeakc0de** [comment]
**Logs**
```
time="2020-07-08T11:26:15Z" level=info msg="[PreReq]: Getting the ENV for the  experiment"
time="2020-07-08T11:26:15Z" level=info msg="[PreReq]: Updating the chaos result of node-cpu-hog experiment (SOT)"
time="2020-07-08T11:26:15Z" level=info msg="The application informations are as follows" Namespace=shubham Label="run=nginx" Chaos Duration=60 Ramp Time=10
time="2020-07-08T11:26:15Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (pre-chaos)"
time="2020-07-08T11:26:15Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-07-08T11:26:15Z" level=info msg="The running status of Pods are as follows" Pod=nginx1-bfb66d6c9-mpbsc Status=Running
time="2020-07-08T11:26:15Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-07-08T11:26:15Z" level=info msg="The running status of container are as follows" container=nginx Pod=nginx1-bfb66d6c9-mpbsc Status=Running
time="2020-07-08T11:26:15Z" level=info msg="[Status]: Verify that the Auxiliary Applications are running (pre-chaos)"
time="2020-07-08T11:26:15Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-07-08T11:26:15Z" level=info msg="The running status of Pods are as follows" Pod=nginx1-bfb66d6c9-mpbsc Status=Running
time="2020-07-08T11:26:15Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-07-08T11:26:15Z" level=info msg="The running status of container are as follows" container=nginx Pod=nginx1-bfb66d6c9-mpbsc Status=Running
time="2020-07-08T11:26:15Z" level=info msg="[Info]: Details of application under chaos injection" NodeCPUcores=2 NodeName=minikube
time="2020-07-08T11:26:15Z" level=info msg="[Ramp]: Waiting for the 10s ramp time before injecting chaos"
time="2020-07-08T11:26:25Z" level=info msg="[Status]: Checking the status of the helper pod"
time="2020-07-08T11:26:25Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-07-08T11:26:33Z" level=info msg="The running status of Pods are as follows" Pod=node-cpu-hog-lrhgeh Status=Running
time="2020-07-08T11:26:33Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-07-08T11:26:33Z" level=info msg="The running status of container are as follows" container=node-cpu-hog Pod=node-cpu-hog-lrhgeh Status=Running
time="2020-07-08T11:26:33Z" level=info msg="[Wait]: Waiting for 90s till the completion of the helper pod"
time="2020-07-08T11:26:33Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:34Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:35Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:36Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:38Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:39Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:40Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:41Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:42Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:43Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:44Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:45Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:46Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:47Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:48Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:49Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:50Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:51Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:52Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:53Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:54Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:55Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:56Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:57Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:58Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:26:59Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:00Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:01Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:02Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:03Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:04Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:05Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:06Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:07Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:08Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:09Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:10Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:11Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:12Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:13Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:14Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:15Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:16Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:17Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:18Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:19Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:20Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:21Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:22Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:23Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:24Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:25Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:26Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:27Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:28Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:30Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:31Z" level=info msg="helper pod status: Running"
time="2020-07-08T11:27:32Z" level=info msg="helper pod status: Succeeded"
time="2020-07-08T11:27:32Z" level=info msg="The running status of Pods are as follows" Pod=node-cpu-hog-lrhgeh Status=Succeeded
time="2020-07-08T11:27:32Z" level=info msg="[Status]: Getting the status of application node"
time="2020-07-08T11:27:32Z" level=info msg="The running status of Nodes are as follows" Node=minikube Ready=true
time="2020-07-08T11:27:32Z" level=info msg="[Cleanup]: Deleting the helper pod"
time="2020-07-08T11:27:32Z" level=info msg="[Ramp]: Waiting for the 10s ramp time after injecting chaos"
time="2020-07-08T11:27:42Z" level=info msg="[Confirmation]: CPU of the application node has been stressed successfully"
time="2020-07-08T11:27:42Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (post-chaos)"
time="2020-07-08T11:27:42Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-07-08T11:27:42Z" level=info msg="The running status of Pods are as follows" Pod=nginx1-bfb66d6c9-mpbsc Status=Running
time="2020-07-08T11:27:42Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-07-08T11:27:42Z" level=info msg="The running status of container are as follows" container=nginx Pod=nginx1-bfb66d6c9-mpbsc Status=Running
time="2020-07-08T11:27:42Z" level=info msg="[Status]: Verify that the Auxiliary Applications are running (post-chaos)"
time="2020-07-08T11:27:42Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-07-08T11:27:42Z" level=info msg="The running status of Pods are as follows" Status=Running Pod=nginx1-bfb66d6c9-mpbsc
time="2020-07-08T11:27:42Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-07-08T11:27:42Z" level=info msg="The running status of container are as follows" Pod=nginx1-bfb66d6c9-mpbsc Status=Running container=nginx
time="2020-07-08T11:27:42Z" level=info msg="[The End]: Updating the chaos result of node-cpu-hog experiment (EOT)"

```

File: chaoslib/litmus/node_cpu_hog/node-cpu-hog.go [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+	log.Infof("[Wait]: Waiting for %vs till the completion of the helper pod", strconv.Itoa(experimentsDetails.ChaosDuration+30))
+
+	err = status.WaitForCompletion(experimentsDetails.ChaosNamespace, "name=node-cpu-hog-"+experimentsDetails.RunID, clients, experimentsDetails.ChaosDuration+30)
+	if err != nil {
+		return err
+	}
+
+	// Checking the status of application node
+	log.Info("[Status]: Getting the status of application node")
+	err = status.CheckNodeStatus(appNodeName, clients)
```
Comment (ksatchit) [inline chaoslib/litmus/node_cpu_hog/node-cpu-hog.go:79]: It is possible we force node in NotReady via cores count.

Comment (ksatchit) [reply chaoslib/litmus/node_cpu_hog/node-cpu-hog.go:79]: Ideally, the check should be to see if CPU utilization has really increased from the initial value we started the experiment with. That is what is indicative of actual chaos.

File: chaoslib/litmus/node_cpu_hog/node-cpu-hog.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+	if err != nil {
+		return errors.Errorf("application node is not in ready state, err: %v", err)
```
Comment (ksatchit) [inline chaoslib/litmus/node_cpu_hog/node-cpu-hog.go:81]: If node not ready, Log it. Ask for recovery to be manual (warning)

File: chaoslib/litmus/node_cpu_hog/node-cpu-hog.go (hunk 1 above)
Comment (uditgaurav) [reply chaoslib/litmus/node_cpu_hog/node-cpu-hog.go:79]: We can try by giving some high manual value like `10000`

File: chaoslib/litmus/node_cpu_hog/node-cpu-hog.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+	err = status.WaitForCompletion(experimentsDetails.ChaosNamespace, "name=node-cpu-hog-"+experimentsDetails.RunID, clients, experimentsDetails.ChaosDuration+30)
+	if err != nil {
+		return err
+	}
+
+	// Checking the status of application node
+	log.Info("[Status]: Getting the status of application node")
+	err = status.CheckNodeStatus(appNodeName, clients)
+	if err != nil {
+		log.Warn("Application node is not in ready state, Please do a manual recovery for this")
```
Comment (ksatchit) [inline chaoslib/litmus/node_cpu_hog/node-cpu-hog.go:81]:
```suggestion
		log.Warn("Application node is not in the ready state, you may need to manually recover the node")
```

It is possible that managed k8s nodes will self-heal also

**Pull request merged**